cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowContext.java

```diff
@@ -51,4 +52,10 @@
    * @return a {@link WorkflowToken}
    */
   WorkflowToken getToken();
+
+  /**
+   * @return the {@link WorkflowActionContext} for custom actions
+   */
+  @Nullable
+  WorkflowActionContext getWorkflowActionContext();
```

Comment:
Datasets can be injected using @UseDataset, but for services you still need to get the context and get the service URL (like in the run method in DatasetWithCustomActionApp)